src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -221,7 +240,11 @@ private static MethodSymbol GetEntryPoint(CSharpCompilation compilation, PEModul
                 moduleBeingBuilt.SetMethodBody(synthesizedEntryPoint, emittedBody);
             }
 
-            Debug.Assert((object)entryPoint != null || entryPointAndDiagnostics.Diagnostics.HasAnyErrors() || !compilation.Options.Errors.IsDefaultOrEmpty);
+            if (addedDefinition && moduleBeingBuilt != null)
+            {
+                moduleBeingBuilt.AddSynthesizedDefinition(entryPoint.ContainingType, synthesizedEntryPoint);
+            }
```

Comment:
For some reason I pulled this out specifically, but that might have been because I was fixing other bugs in the implementation and wanted to isolate this.

Changing it back to what you recommend does work.